# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data['children'].value_counts() # Вспомним, какие варианты количества детей представлены в таблице.

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

В таблице представлено несколько категорий семей в зависимости от количества детей. Больше всего семей без детей - около 14 000. Затем идут семьи с одним ребенком - около 4800. На третьем месте - семьи с двумя детьми - около 2000. Меньше всего многодетных семей. Возможно, имеет смысл объединить их в одну категорию. Чтобы проверить это, мы построим две сводные таблицы. Первая будет построена по категориям "бездетная семья", "среднестатистическая семья", "многодетная семья", а вторая - по количеству детей.

In [23]:
# Для удобства анализа я разделю семьи на три категории с помощью функции создания условий if.
def children_category_four(children):
    if children == 0:
        return 'Бездетная семья'
    if children == 1 or children == 2:
        return 'Среднестатистическая семья'
    if children >= 3:
        return 'Многодетная семья'
    
# Создам новый столбец в нашей таблице и добавлю в него данные, используя функцию обработки данных из столбца children.
data['children_category'] = data['children'].apply(children_category_four)

# Проверю правильность работы функции, выведя на экран два столбца таблицы.
data[['children','children_category']].head(10)

children           children_category
0         1  Среднестатистическая семья
1         1  Среднестатистическая семья
2         0             Бездетная семья
3         3           Многодетная семья
4         0             Бездетная семья
5         0             Бездетная семья
6         0             Бездетная семья
7         0             Бездетная семья
8         2  Среднестатистическая семья
9         0             Бездетная семья

In [24]:
# Введу функцию, которая будет обрабатывать данные из столбца с фактами задолженности.
def per_debt(pdSerises):
    return str(round((pdSerises.sum() / pdSerises.count()) * 100, 2)) + '%'

# Создам сводную таблицу с двумя столбцами - количеством заемщиков определенной категории и % имевших задолженностей среди этих категорий.
data_pivot = data.pivot_table(index=['children'], values=['debt'], aggfunc=['count',per_debt])

# Выведу результат отсориториванной сводной таблицы, с теми группами заемщиков, которых больше всего.
data_pivot[data_pivot['count', 'debt'] > 1000].sort_values(by=('per_debt', 'debt'))


count per_debt
           debt     debt
children                
0         14091    7.54%
1          4808    9.23%
2          2052    9.45%

Система классификации семей работает правильно. Теперь нам нужно обработать данные в столбце с количеством просрочек. Для этого нам понадобится использовать информацию из столбца 'debt'. Согласно документу, 'debt' указывает, была ли у клиента задолженность по возврату кредитов. Посмотрим, какие уникальные значения присутствуют в этом столбце и какого типа эти данные.

In [25]:
data['debt'].value_counts()

0    19599
1     1732
Name: debt, dtype: int64

Теперь нам потребуется обработать эти данные, чтобы таблица отражала % просрочек в зависимости от категории семьи или количества детей.

In [26]:
# Создаю вторую функцию per_debt, которая будет выводить нам процент просрочек.
def per_debt(pdSerises):
    return str(round((pdSerises.sum() / pdSerises.count()) * 100, 2)) + '%'

In [27]:
# Построю сводную таблицу data_pivot с помощью инструмена pivot_table. Установлю фильтр по категории семьи. В качестве значений возьму столбец просрочек debt, обработаю его с помощью 3-х функций.
data_pivot = data.pivot_table(index=['children_category'], values=['debt'], aggfunc=['sum','count',per_debt])

# Выведу результат отсориториванной сводной таблицы.
data_pivot.sort_values(by=('per_debt', 'debt'))

sum  count per_debt
                            debt   debt     debt
children_category                               
Бездетная семья             1063  14091    7.54%
Многодетная семья             31    380    8.16%
Среднестатистическая семья   638   6860     9.3%

В первой сводной таблице мы видим, что семьи без детей имеют наименьшее количество фактов задолженности по возврату кредитов. Среднестатистические семьи с 1-2 детьми, вероятно, не являются самыми надежными заемщиками согласно первой сводной таблице. Чтобы подтвердить этот вывод, мы создадим еще одну сводную таблицу, не от категории семьи, а от количества детей. Возможно, семьи с одним ребенком будут наравне с бездетными, а после появления второго ребенка количество просрочек резко возрастет.

In [28]:
# Построю сводную таблицу data_pivot с помощью инструмена pivot_table. Установлю фильтр по количеству детей в семье. В качестве значений возьму столбец просрочек debt. Выведу столбцы, обработанные тремя функциями.
data_pivot = data.pivot_table(index=['children'], values=['debt'], aggfunc=['sum','count',per_debt])

# Далее выведу результат второй сводной таблицы.
data_pivot.sort_values(by=('per_debt', 'debt'))

sum  count per_debt
          debt   debt     debt
children                      
5            0      9     0.0%
0         1063  14091    7.54%
3           27    330    8.18%
1          444   4808    9.23%
2          194   2052    9.45%
4            4     41    9.76%

Выводы из первой сводной таблицы о том, что семьи с одним ребенком такие же надежные, как и бездетные семьи, не подтвердилась. Процент фактов задолженностей семей с одним ребенком незначительно отличается от результатов этого же критерия для семей с двумя и четырьмя детьми.
Однако, в текущей выборке мы обнаружили, что самыми ответственными являются семьи с пятью детьми, хотя они и являются наименьшими представителями по количеству таких семей в нашей таблице.

**Вывод:** 
В анализируемой таблице большинство семей - бездетные (около 14 000), у которых наименьший процент возникновения задолженности (7,5%). Семьи с тремя детьми чуть менее надежны, у них процент возникновения задолженности составляет 8,2%. Остальные семьи не сильно отличаются по проценту возникновения задолженностей (от 9,2% до 9,7%). Однако, результаты оценки надежности плательщиков в зависимости от количества детей не являются однозначными, так как разница между самыми надежными и ненадежными плательщиками составляет всего 2,2%. Вероятно, этот метод оценки надежности не является достаточно надежным. Необходимо проверить другие методы оценки надежности и сравнить разницу между самым низким и самым высоким показателем надежности в каждом методе.

**Вывод №2:**  Семьи с пятью детьми имеют наименьший процент задолженностей, однако их количество в выборке крайне мало (всего 9 семей), поэтому этот результат не может быть полностью репрезентативным.

Бездетные семьи составляют большинство в выборке (около 14 000) и имеют наименьший процент задолженностей (около 7,5%). Семьи с тремя детьми немного менее надежны, у них процент задолженностей составляет около 8,2%. Остальные семьи имеют схожие показатели задолженностей, в диапазоне от 9,2% до 9,7%.

Оценка надежности плательщиков по количеству детей не является однозначной, так как разница между самыми надежными и ненадежными плательщиками составляет всего 2,2%. Этот метод оценки не рекомендуется использовать как основной, но его следует проверить другими способами и сравнить результаты.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Проведем аналогичные операции с построением таблицы, но в данном случае нам не нужно обрабатывать данные столбца под %. Сначала мы посмотрим, людей с каким семейным положением больше, чтобы определить, на какие данные можно опираться, а какие не принимать во внимание.

In [29]:
data['family_status'].value_counts()

женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64

В таблице больше всего людей, находящихся в зарегистрированном браке (около 12 000), в три раза меньше заемщиков в гражданском браке (около 4000), заемщиков в статусе не женат/не замужем (около 3000). Количество заемщиков, находящихся в разводе и овдовевших, примерно одинаково (951 и 1189 соответственно). Можно разделить заемщиков на занятых и холостых, но это не имеет особого смысла на данный момент.

Таким образом, можно учитывать все категории заемщиков, так как здесь не наблюдается аномально низких результатов, как было в первом случае с категориями семей.

In [30]:
# Построю сводную таблицу с помощью инструмена pivot_table. Установлю фильтр по семейному положению. В качестве значений возьму столбец просрочек debt. Выведу столбцы, обработанные тремя функциями.
data_pivot_family_status = data.pivot_table(index=['family_status'], values=['debt'], aggfunc=['sum','count',per_debt])

# Выведу результат.
data_pivot_family_status.sort_values(by=('per_debt', 'debt'))

sum  count per_debt
                      debt   debt     debt
family_status                             
вдовец / вдова          63    951    6.62%
в разводе               84   1189    7.06%
женат / замужем        927  12261    7.56%
гражданский брак       385   4134    9.31%
Не женат / не замужем  273   2796    9.76%

**Вывод:** Мы видим, что самыми надежными заемщиками являются овдовевшие заемщики, но их меньше всего в нашей таблице. Чуть менее надежными заемщиками являются те, кто находится в разводе. На третьем месте по надежности выплат находятся заемщики, которые находятся сейчас в зарегистрированном браке. Их надежность примерно равна результатам надежности бездетных заемщиков. Также мы видим, что представителей замужних/женатых больше всего. Самыми ненадежными заемщиками являются находящиеся в гражданском браке заемщики. Заемщики, хоть раз заключавшие брак (находящиеся в нем или уже нет), являются наиболее надежными. Результаты оценки надежности плательщиков в зависимости от семейного положения более показательны, чем результаты оценки надежности по количеству детей. Результаты более показательные, но не совсем однозначные.



#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Ранее мы уже создавали столбец ['total_income_category'] с категоризацией заемщиков по уровню дохода. Эти категории были следующими:

0–30000 — 'E';

30001–50000 — 'D';

50001–200000 — 'C';

200001–1000000 — 'B';

1000001 и выше — 'A'


Посмотрим какой категории заемщиков у нас больше всего.

In [31]:
data['total_income_category'].value_counts()

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64

Наибольшее количество заемщиков относится к категории С, их около 16000 человек. Заемщиков категории В, имеющих больший доход, примерно в три раза меньше - чуть более 5000 человек. Количество заемщиков с другими категориями составляет менее 500 человек в общей сложности.

In [32]:
# Построю сводную таблицу с помощью инструмена pivot_table. Установлю фильтр по категории дохода. В качестве значений возьму столбец просрочек debt. Выведу столбцы, обработанные тремя функциями.
data_pivot_income_category = data.pivot_table(index=['total_income_category'], values=['debt'], aggfunc=['sum','count',per_debt])

# Выведу результат.
data_pivot_income_category.sort_values(by=('per_debt', 'debt'))

sum  count per_debt
                       debt   debt     debt
total_income_category                      
D                        21    349    6.02%
B                       354   5014    7.06%
A                         2     25     8.0%
C                      1353  15921     8.5%
E                         2     22    9.09%

**Вывод:** Согласно анализу, самые надежные заемщики относятся к категории D, у которых доход составляет от 30 001 до 50 000 рублей. Однако таких заемщиков немного - всего 349 человек. Следующими по надежности являются заемщики категории B, имеющие высокий доход от 200 001 до 1 000 000 рублей. Их количество достаточно велико, чтобы учитывать эти данные при принятии решений.

Заемщики с очень высоким доходом (более миллиона рублей) не являются самыми надежными, но и не относятся к наименее надежным. Они занимают среднее положение по надежности.

Самыми ненадежными заемщиками являются те, кто относится к категориям E (с самым низким доходом до 30 000 рублей) и C (со средним доходом от 50 001 до 200 000 рублей). Категория C является наиболее распространенной среди заемщиков.

Оценка надежности плательщиков на основе уровня дохода дает результаты, аналогичные оценке на основе семейного положения. Разница между самыми надежными и ненадежными плательщиками составляет примерно 3,5%. Эти результаты более показательны и однозначны.

In [33]:
# Построю сводную таблицу с помощью инструмена pivot_table для всех категорий заемщиков.
data_pivot_income_category = data.pivot_table(index=['total_income_category'], values=['debt'], aggfunc=['count',per_debt])

# Выведу результат сводной таблицы, с преобладающими категориями заемщиков.
data_pivot_income_category[data_pivot_income_category['count', 'debt'] > 5000].sort_values(by=('per_debt', 'debt'))

count per_debt
                        debt     debt
total_income_category                
B                       5014    7.06%
C                      15921     8.5%

В данной таблице представлены данные о количестве заемщиков и проценте задолженности по категориям дохода. Категория B включает 5014 заемщиков, из которых 7,06% имеют задолженность. Категория C включает 15921 заемщика, из которых 8,5% имеют задолженность.

Категория B включает заемщиков с достаточно высоким доходом, от 200 001 до 1 000 000 рублей. Представителей этой группы заемщиков достаточно много, чтобы учитывать их данные при принятии решений.

Категория C включает заемщиков со средним уровнем дохода, от 50 001 до 200 000 рублей. Эта группа является наиболее распространенной среди заемщиков. Эти результаты являются показательными на мой взгляд.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Ранее мы также создавали столбец ['purpose_category'] который классифицирует заемщиков по цели получения кредита. Посмотрим на эти категории и подсчитаем, сколько заемщиков каждой категории у нас есть в данных.

In [34]:
data['purpose_category'].value_counts()

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

Мы видим, что чаще всего берут кредит на покупку недвижимости. Примерно на одном уровне находятся и другие категории, число представителей которых достигает тысяч.

In [35]:
# Построю сводную таблицу с помощью инструмена pivot_table. Установлю фильтр по целям займа. В качестве значений возьму столбец просрочек debt. Выведу столбцы, обработанные тремя функциями.
data_pivot_purpose_category = data.pivot_table(index=['purpose_category'], values=['debt'], aggfunc=['sum','count',per_debt])

# Выведу результат.
data_pivot_purpose_category.sort_values(by=('per_debt', 'debt'))

sum  count per_debt
                         debt   debt     debt
purpose_category                             
операции с недвижимостью  780  10751    7.26%
проведение свадьбы        183   2313    7.91%
получение образования     369   3988    9.25%
операции с автомобилем    400   4279    9.35%

**Вывод:** Заемщики, которые берут кредит на операции с недвижимостью, являются наиболее надежными и составляют наибольшую категорию. Это делает их самым безопасным вариантом для банка. Заемщики, берущие кредит на проведение свадьбы, также считаются относительно надежными, что подтверждается проверкой других гипотез, согласно которым женатые/замужние или разведенные люди являются надежными заемщиками.
Наименее ответственными и надежными заемщиками являются те, кто берет кредит на образование и приобретение автомобиля.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* В этой таблице есть пустые ячейки в столбцах "days_employed" и "total_income". Вероятно, это связано с тем, что некоторые люди не хотели отвечать на эти вопросы. Интересно, что количество пропущенных значений в обоих столбцах одинаково. Это может говорить о том, что заполнение одного из этих столбцов влияет на заполнение другого. 

Также причинами появления пропусков могут быть:
1. Технические проблемы - данные могут быть потеряны или повреждены из-за технических сбоев.
2. Неправильное заполнение формы - некоторые люди могут неправильно заполнить форму.
3. Отсутствие информации - в некоторых случаях информация может просто отсутствовать.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медиана - это такое значение, которое делит выборку на две равные части, одна половина значений выше медианы, другая - ниже. Она помогает избежать влияния аномально высоких или низких значений на среднее значение. Медиана является лучшим показателем для оценки выборки в целом, особенно, когда есть незаполненные данные.

### Шаг 4: общий вывод.

При оценке надежности заемщиков по различным критериям, были получены следующие результаты:

По категории семей: Бездетные семьи показали самую высокую надежность, с долей просроченных займов 7.54%.

По семейному положению заемщиков: Вдовцы/вдовы, разведенные и вступившие в брак заемщики показали высокую надежность, с долей просроченных займов менее 7.6%.

По категории доходности заемщиков: Самыми надежными оказались заемщики категорий D и B, с долей просроченных займов менее 7.06%.

По целям кредита: Самыми надежными клиентами оказались те, кто приобретает недвижимость на заемные средства, с долей просроченных займов менее 7.26%.


Вторая интерпритация: 
В рамках данного проекта я провела исследование, чтобы определить, влияет ли семейное положение и количество детей клиента на своевременный возврат кредита. Я проанализировала, существует ли связь между количеством детей и своевременным погашением кредита, есть ли связь между семейным положением и своевременным возвратом кредита, есть ли зависимость между уровнем дохода и своевременным погашением кредита и как различные цели кредитования влияют на его своевременный возврат.

Бездетные семьи показали наивысшую надежность при оценке по категории детности семей, имея долю просроченных займов в размере 7,54%. Вдовцы, разведенные и вступившие в брак заемщики продемонстрировали высокую надежность при оценке по семейному положению, с долей просроченных займов менее 7,6%. Заемщики категорий D и B оказались самыми надежными при оценке по категории доходности, с долей просроченных займов менее 7,06%. Те, кто приобретает недвижимость за заемные средства, оказались самыми надежными клиентами при оценке по целям кредита, с долей просроченных займов менее 7,26%.

Если оценивать по двум критериям - доходности и семейному положению - наиболее надежными оказались заемщики, заключившие брак, но с низкой доходностью (30-50 тыс.), и вдовцы/вдовы с высокой доходностью. Вероятность просрочки у таких заемщиков составляет менее 7,1%.

На мой взгляд, оценка финансовой надежности заемщиков по одному критерию не всегда дает надежные результаты, так как все четыре метода оценки по одному критерию показывают примерно одинаковую дельту в результатах. 